### PR TITLE
Provide support for Azure Key Vault in TriggerAuthentication.

### LINKED ISSUE #900: Provide Azure Key Vault support in TriggerAuthentication
Provide Azure Key Vault support in TriggerAuthentication to pull secrets from.

### PR DESCRIPTION
**Pull request by v-shenoy**
Authentication via Azure Key Vault is now supported.  
Sample for the new TriggerAuthenticationSpec - 

```yaml
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: triggerAuthName
  namespace: default
spec:
  azureKeyVault:
    vaultUri: <vault-address>
    credentials:  
      clientId: <azureAD-clientID>
      clientSecret:
        valueFrom:
          secretKeyRef:
            name: <secret-name>
            key: <key-within-secret>
      tenantId: <azureAD-tenantID>
    secrets: 
    - parameter: <param-name-for-authenticating>
      name: <secret-name-in-key-vault>
      version: <secret-version> # Optional

```

Edit 1 - Updated spec and checklist.
Edit 2- Raised documentation PR.
Edit 3 - Update checklist with tests added.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes  #900

### CODE REVIEW

**Comment by tomkerkhove** [comment]
Can we also support managed identity or not?

**Comment by tomkerkhove** [comment]
> ```yaml
> apiVersion: keda.sh/v1alpha1
> kind: TriggerAuthentication
> metadata:
>   name: triggerAuthName
>   namespace: default
> spec:
>   azureKeyVault:
>     vaultUri: <vault-address>
>     credentials:  
>       clientId: <azureAD-clientID>
>       clientSecret:
>         kubeSecret: <kubernetes-secret-containing-azureAD-secret>
>         key: <key-within-the-kubernetes-secret>
>       tenantId: <azureAD-tenantID>
>     secrets: 
>     - parameter: <param-name-for-authenticating>
>       name: <secret-name-in-key-vault>
>       version: <secret-version> # Optional
> ```

Some quick feedback on `azureKeyVault.credentials.clientSecret`. Can we align with the existing secret ref approach in Kubernetes?

**Comment by v-shenoy** [comment]
> Can we also support managed identity or not?

I am not really sure how to go about this. Need some guidance.

**Comment by tomkerkhove** [comment]
The native Kubernetes approach - https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables:

```yaml
valueFrom:
  secretKeyRef:
    name: mysecret
    key: username
```

**Comment by v-shenoy** [comment]
> The native Kubernetes approach - https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables:
> 
> ```yaml
> valueFrom:
>   secretKeyRef:
>     name: mysecret
>     key: username
> ```

So, when you say align with it, you mean just renaming the object to use `valueFrom`, `secretKeyRef` instead of `clientSecret`?

**Comment by tomkerkhove** [comment]
I would go with this:
```yaml
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: triggerAuthName
  namespace: default
spec:
  azureKeyVault:
    vaultUri: <vault-address>
    credentials:  
      clientId: <azureAD-clientID>
      clientSecret:
        valueFrom:
          secretKeyRef:
            name: mysecret
            key: username
      tenantId: <azureAD-tenantID>
    secrets: 
    - parameter: <param-name-for-authenticating>
      name: <secret-name-in-key-vault>
      version: <secret-version> # Optional
```

**Comment by v-shenoy** [comment]
> I would go with this:
> 
> ```yaml
> apiVersion: keda.sh/v1alpha1
> kind: TriggerAuthentication
> metadata:
>   name: triggerAuthName
>   namespace: default
> spec:
>   azureKeyVault:
>     vaultUri: <vault-address>
>     credentials:  
>       clientId: <azureAD-clientID>
>       clientSecret:
>         valueFrom:
>           secretKeyRef:
>             name: mysecret
>             key: username
>       tenantId: <azureAD-tenantID>
>     secrets: 
>     - parameter: <param-name-for-authenticating>
>       name: <secret-name-in-key-vault>
>       version: <secret-version> # Optional
> ```

I'll make the changes.

**Comment by tomkerkhove** [comment]
Since this is our data plane I don't think we need any

**Comment by v-shenoy** [comment]
What about these, [02-crd-clustertriggerauthentications.yaml](https://github.com/kedacore/charts/blob/main/keda/templates/02-crd-clustertriggerauthentications.yaml), [05-crd-triggerauthentications.keda.sh.yaml](https://github.com/kedacore/charts/blob/main/keda/templates/05-crd-triggerauthentications.keda.sh.yaml)?

**Comment by tomkerkhove** [comment]
CRDs is a different story but I'll need @zroubalik for this.

**Comment by tomkerkhove** [comment]
Also, would you mind opening a documentation PR for this nice addition please?

File: pkg/scaling/resolver/azure_keyvault_handler.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (tomkerkhove) [inline pkg/scaling/resolver/azure_keyvault_handler.go:1]: I'm not a Go expert but let's make sure we surface throttling errors from Azure Key Vault since it has low service limits.

Ideally we would also:
- Cache secret for a few minutes
- Re-fetch cached secret if it was unauthorized to see if it was rotated or not.

**Comment by v-shenoy** [comment]
> Also, would you mind opening a documentation PR for this nice addition please?

Yeah, I am in the middle of doing it.

File: pkg/scaling/resolver/azure_keyvault_handler.go (hunk 1 above)
Comment (v-shenoy) [reply pkg/scaling/resolver/azure_keyvault_handler.go:1]: Currently doing the documentation changes, will start on this right after.

**Comment by v-shenoy** [comment]
As the helm charts need to be updated due to the CRD changes, what version, appVersion do I use [here](https://github.com/kedacore/charts/blob/main/keda/Chart.yaml#L11)?

**Comment by v-shenoy** [comment]
Raised a PR for the documentation - https://github.com/kedacore/keda-docs/pull/704

**Comment by JorTurFer** [comment]
I'm not an expert in Azure Key Vault, but the code looks good.
Could you add  an e2e test that uses the Key Vault? 
Can we provide it @tomkerkhove?

**Comment by tomkerkhove** [comment]
Certainly, we can store a Service Bus connection string in the vault to ensure it can authenticate to it. That way, we can re-use existing SB tests - What do you think?

**Comment by JorTurFer** [comment]
I'd do another test only for checking this. AFAIR, if the connection string fails, KEDA doesn't create the HPA, we could store one connection string on KeyVault and try to use it. Or maybe doing the query manually yas for checking that there is not error.

But I'd go with another test

**Comment by zroubalik** [comment]
> I'd do another test only for checking this. AFAIR, if the connection string fails, KEDA doesn't create the HPA, we could store one connection string on KeyVault and try to use it. Or maybe doing the query manually just for checking that response is not an error.
> 
> But I'd go with another test

Yeah, we should definitely add a standalone test for this -> but nothing prevents us to duplicate the existing test and extend it with Vault usage :)

**Comment by tomkerkhove** [comment]
Agreed

**Comment by v-shenoy** [comment]
I can't find an E2E test for ServiceBus [here](https://github.com/kedacore/keda/tree/main/tests/scalers). Should I duplicate another scaler test to use with the key vault?

**Comment by zroubalik** [comment]
> I can't find an E2E test for ServiceBus [here](https://github.com/kedacore/keda/tree/main/tests/scalers). Should I duplicate another scaler test to use with the key vault? And do I add them in the `tests/scalers/` folder itself?

Ah right, the very best option would be to add a ServiceBus E2E test, do you think you can contribute this? And then add a extended version of this e2e test that will use the Azure Key Vault.

**Comment by JorTurFer** [comment]
yeah! if you add also service bus tests, it would be nice!!
But if you prefer not doing it, you can use other already existing test which needs auth, I mean, the idea is recovering the values from the vault, but it's not important what door they open IMO

**Comment by v-shenoy** [comment]
I am not too familiar with TypeScript or the ava framework, so I'll take a look and see if I can add the ServiceBus test. If not I was planning on using the  existing `azure-queue-trigger-auth.test`. And also add a `azure_key_vault_handler_test` unit test.

**Comment by tomkerkhove** [comment]
Let's do this:
- add unit tests in this PR
- add e2e tests for key vault in this PR
- add e2e test using key vault with Azure storage scaler
- I will open issue for service bus e2e tests as follow-up by somebody having time

**Comment by v-shenoy** [comment]
I don't think it makes sense to have unit test considering the `AzureKeyVaultHandler` is only initialized and fetches secrets. Have added the e2e test using queue trigger. Please take a look. @tomkerkhove @JorTurFer @zroubalik

**Comment by tomkerkhove** [comment]
> I don't think it makes sense to have unit test considering the `AzureKeyVaultHandler` is only initialized and just fetches secrets.

I don't think this can hurt, no?

**Comment by v-shenoy** [comment]
The struct exposes only two methods, Initialize, Read. Not really sure how to test those without actually providing actual uri, creds and try to fetch a secret. But that would not be a unit test.

**Comment by tomkerkhove** [comment]
Fine by me if it's fine for the others

File: tests/.env [hunk 2]
```diff
@@ -16,3 +16,7 @@ AZURE_DEVOPS_PAT=
 AZURE_DEVOPS_PROJECT=
 AZURE_DEVOPS_BUILD_DEFINITON_ID=
 AZURE_DEVOPS_POOL_NAME=
+AZURE_KEYVAULT_URI=
+AZURE_AD_CLIENT_ID=
+AZURE_AD_CLIENT_SECRET=
+AZURE_AD_TENANT_ID=
```
Comment (zroubalik) [inline tests/.env:22]: @tomkerkhove can we have these in GH secrets?

File: pkg/scaling/resolver/azure_keyvault_handler.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+
+	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
```
Comment (zroubalik) [inline pkg/scaling/resolver/azure_keyvault_handler.go:28]: Just a nit, we tend to have general imports in one block and KEDA related imports in a new block right below this block.
```suggestion
	"github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault"
	"github.com/Azure/go-autorest/autorest/azure/auth"
	"github.com/go-logr/logr"
	"sigs.k8s.io/controller-runtime/pkg/client"

	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -32,6 +32,7 @@
 
 - **General:** Automatically release container image for ARM ([#2263]https://github.com/kedacore/keda/issues/2263))
 - **General:** Automatically run end-to-end tests on ARM ([#2262]https://github.com/kedacore/keda/issues/2262))
+- **General:** Provide support for authentication via Azure Key Vault  ([#900](https://github.com/kedacore/keda/issues/900))
```
Comment (zroubalik) [inline CHANGELOG.md:35]: Sorry for being pain in the ***, but could you please rebase and keep alphabetical order here? :)

**Review by zroubalik — COMMENTED** [review]
Looking good, just some minor nits

File: tests/.env (hunk 2 above)
Comment (v-shenoy) [reply tests/.env:22]: Steps for E2E tests to work -

1. Provision a key vault using the subscription used for testing.
2. Register an application with Azure AD, and give the "read secrets" permission to it on the keyvault.
3. Add the keyvault uri, and AD credentials as GitHub secrets.

Comment (zroubalik) [reply tests/.env:22]: By chance, is this possible to do via API as part of the test itself?

Comment (v-shenoy) [reply tests/.env:22]: I could try to do this by running shell commands from the test using the `az cli` but I don't think that's clean. And I don't know about the library to use for providing RBAC / access policy to app on the keyvault.

Comment (tomkerkhove) [reply tests/.env:22]: I'll do it, it's OK.

Comment (tomkerkhove) [reply tests/.env:22]: Please change to use these instead please:
- `AZURE_SP_ID`
- `AZURE_SP_KEY`
- `AZURE_SP_TENANT`

File: tests/scalers/azure-keyvault-queue.test.ts [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+      clientId: ${azureADClientID}
+      tenantId: ${azureADTenantID}
+      clientSecret:
+        valueFrom:
+          secretKeyRef:
+            name: test-auth-secrets
+            key: clientSecret
+    secrets:
+    - parameter: connection
+      name: storageConnectionString
```
Comment (tomkerkhove) [inline tests/scalers/azure-keyvault-queue.test.ts:164]: Please use `E2E-Storage-ConnectionString` instead as secret name
```suggestion
      name: E2E-Storage-ConnectionString
```

File: tests/.env (hunk 2 above)
Comment (tomkerkhove) [reply tests/.env:22]: https://github.com/kedacore/keda/pull/2756

**Review by zroubalik — APPROVED** [review]
LGTM

**Comment by SebastianSchuetze** [comment]
Last question for me. Does this support also managed identity as requested or does this come later?

**Comment by v-shenoy** [comment]
> Last question for me. Does this support also managed identity as requested or does this come later?

It doesn't currently, it will have to be taken up later.

**Comment by mingue** [comment]
Looking into using this once released, but will this also be supported for ClusterTriggerAuthentication? The updated documentation doesn't make any exception. If so where will it get the credentials for the clientSecret:

```yaml
clientSecret:
  valueFrom:
    secretKeyRef: # Where is this secret coming from in a ClusterTriggerAuthentication?
      name: <secret-name>
      key: <key-within-secret>
```

**Comment by tomkerkhove** [comment]
That's something we should improve in our docs indeed, @v-shenoy I presume this would be from a secret in the same namespace as the ClusterTriggerAuthentication itself, right?

**Comment by v-shenoy** [comment]
Yes, it would be. Do we add this to the documentation, or should we have cluster trigger authentication be able to fetch secrets from other namespaces as well by specifying a namespace key in the object as well? (Only for Cluster, not normal trigger auth)

**Comment by tomkerkhove** [comment]
No, I think the current approach is perfect but let's improve the docs a bit.

**Comment by v-shenoy** [comment]
> Last question for me. Does this support also managed identity as requested or does this come later?

While this thread is active, @SebastianSchuetze Pod Identity support is being added to key vault, along with workload identity.

**Comment by tomkerkhove** [comment]
Merged, thanks!
